src/decoder/symphonia.rs

```diff
@@ -273,72 +282,60 @@ impl std::error::Error for SeekError {
 impl SymphoniaDecoder {
     /// Note span offset must be set after
     fn refine_position(&mut self, seek_res: SeekedTo) -> Result<(), source::SeekError> {
-        let mut samples_to_pass = seek_res.required_ts - seek_res.actual_ts;
-        let packet = loop {
-            let candidate = self.format.next_packet().map_err(SeekError::Refining)?;
-            if candidate.dur() > samples_to_pass {
-                break candidate;
-            } else {
-                samples_to_pass -= candidate.dur();
-            }
-        };
-
-        let mut decoded = self.decoder.decode(&packet);
-        for _ in 0..MAX_DECODE_RETRIES {
-            if decoded.is_err() {
-                let packet = self.format.next_packet().map_err(SeekError::Retrying)?;
-                decoded = self.decoder.decode(&packet);
-            }
+        // Calculate the number of samples to skip.
+        let mut samples_to_skip = (Duration::from(
+            self.time_base
+                .ok_or(source::SeekError::SymphoniaDecoder(SeekError::Refining(
+                    "time base not available",
+                )))?
+                .calc_time(seek_res.required_ts.saturating_sub(seek_res.actual_ts)),
+        )
+        .as_secs_f32()
+            * self.sample_rate() as f32
+            * self.channels() as f32)
+            .ceil() as usize;
+
+        // Re-align the seek position to the first channel.
+        samples_to_skip -= samples_to_skip % self.channels() as usize;
+
+        // Skip ahead to the precise position.
+        for _ in 0..samples_to_skip {
+            self.next();
         }
 
-        let decoded = decoded.map_err(SeekError::Decoding)?;
-        decoded.spec().clone_into(&mut self.spec);
-        self.buffer = SymphoniaDecoder::get_buffer(decoded, &self.spec);
-        self.current_span_offset = samples_to_pass as usize * self.channels() as usize;
         Ok(())
     }
 }
 
-fn skip_back_a_tiny_bit(
-    Time {
-        mut seconds,
-        mut frac,
-    }: Time,
-) -> Time {
-    frac -= 0.0001;
-    if frac < 0.0 {
-        seconds = seconds.saturating_sub(1);
-        frac = 1.0 - frac;
-    }
-    Time { seconds, frac }
-}
-
-fn time_to_duration(time: Time) -> Duration {
-    Duration::new(
-        time.seconds,
-        if time.frac > 0.0 {
-            (1f64 / time.frac) as u32
-        } else {
-            0
-        },
-    )
-}
-
 impl Iterator for SymphoniaDecoder {
     type Item = DecoderSample;
 
-    #[inline]
     fn next(&mut self) -> Option<Self::Item> {
         if self.current_span_offset >= self.buffer.len() {
-            let packet = self.format.next_packet().ok()?;
-            let mut decoded = self.decoder.decode(&packet);
-            for _ in 0..MAX_DECODE_RETRIES {
+            let mut decode_errors = 0;
+            let decoded = loop {
+                let packet = self.format.next_packet().ok()?;
```

Comment:
just to make sure this is on your radar (for the future, not this pr). `next_packet` could return `ResetRequired`: 

> then the track list must be re-examined and all `Decoder`s re-created.

Reply:
Yes, on the radar. Currently only occurs on chained Ogg streams. Relevant test files are available at https://people.xiph.org/~xiphmont/test-vectors/vorbis/ - could we include them? Doesn't state the license and then you could debate whether they are defined as "software" under the [Xiph License](https://xiph.org/flac/license.html) or not.

Reply:
go ahead and add them, I'll shoot them a mail. If they tell me its not okay we can remove them.